qutip/core/data/csr.pxd

```diff
@@ -103,7 +104,10 @@ cdef inline base.idxint acc_gather(Accumulator *acc, double complex *values, bas
     for i in range(acc.nnz):
         position = acc.nonzero[i]
         value = acc.values[position]
-        if value != 0:
+        if (
+            value != 0 and
+            (tol == 0 or fabs(value.real) >= tol or fabs(value.imag) >= tol)
+        ):
```

Comment:
why not
```cython
value = acc.values[position]
if fabs(value.real) < tol:
    value.real = 0
if fabs(value.imag) < tol:
    value.imag = 0
if value != 0:
    ...
```
(or alternatively the same way it's currently in `tidyup_csr`).  The way it's written in the PR would allow a value to pass through unaltered if _either_ of its real and imaginary parts are above the tolerance, but that's inconsistent with the behaviour of `tidyup_csr`.

Reply:
Yes, but do we want the current `tidyup` behaviour?
If the aim is to remove small value to keep the matrix sparse, there is no point in cleaning the `real` and `imag` part individually. We only introduce extra numerical error in the operation.
The present `tidyup` help to have nice looking output: `1` looks better than `1+1e-15j`, but other than the look, I don't understand why we round the real and imaginary part independently...

And now that `tidyup` in a method of `Qobj` explicitly called by the user, do `CSR` operations need to be consistent with it? I am tempted to have the `auto_tidyup_atol` and default `tidyup` tolerance being different values and rename the options to make those 2 distinct.

Reply:
I could see the full `tidyup` going either way, but at least for the purposes of this PR, which isn't changing `tidyup_csr`, the methods should be the same - it would be surprising if not, I think.

An argument in favour of keeping it separate for `real` and `imag` is that it makes the `atol` argument have a proper meaning; if you allow a value if _either_ is above the threshold because you're considering the complex number as a whole, then the true maximum absolute value that can get tidied is `sqrt(2) * atol`, and there are numbers with an absolute value less than that which wouldn't get tidied up (`1.1*atol`, for example).  If you want the absolute value to work properly, you end up with 3 more flops per nonzero (though you save one comparison).  The current method is faster; you can also do it with one CPU-vectorised pass over the data reinterpreted as double-precision floats, where each operation is a conditional move (`cmp [x] y; mov 0 [x]`, but one CPU instruction), which (if I remember correctly) are typically immune to branch mispredication.